File: frontend/src/lib/staffPermissions.ts
/**
 * Frontend mirror of backend/app/services/staff_permissions.py.
 * Keep in sync with the backend matrix whenever roles or nav IDs change.
 */
import type { StaffNavId } from '@/components/staff/staffNav';
import type { ActionType } from '@/types';

export type StaffRole =
  | 'manager'
  | 'front_desk'
  | 'maintenance'
  | 'housekeeping'
  | 'room_service';

const ALL_ACTION_TYPES: ReadonlySet<ActionType> = new Set<ActionType>([
  'MAINTENANCE',
  'ROOM_SERVICE',
  'HOUSEKEEPING',
  'CONTACT_FRONT_DESK',
  'HANDOFF',
]);

export const ROLE_NAV: Record<StaffRole, ReadonlySet<StaffNavId>> = {
  manager: new Set<StaffNavId>(['tasks', 'assigned', 'schedule', 'review', 'guest-chat', 'help-desk', 'knowledge']),
  front_desk: new Set<StaffNavId>(['tasks', 'assigned', 'schedule', 'review', 'guest-chat', 'help-desk', 'knowledge']),
  maintenance: new Set<StaffNavId>(['tasks', 'assigned', 'schedule']),
  housekeeping: new Set<StaffNavId>(['tasks', 'assigned', 'schedule']),
  room_service: new Set<StaffNavId>(['tasks', 'assigned', 'schedule']),
};

export const ROLE_ACTION_TYPES: Record<StaffRole, ReadonlySet<ActionType>> = {
  manager: ALL_ACTION_TYPES,
  front_desk: ALL_ACTION_TYPES,
  maintenance: new Set<ActionType>(['MAINTENANCE', 'HANDOFF']),
  housekeeping: new Set<ActionType>(['HOUSEKEEPING', 'HANDOFF']),
  room_service: new Set<ActionType>(['ROOM_SERVICE', 'HANDOFF']),
};

/** Roles that may see the "Get help with this task" button (consumed by Agent 6). */
export const TASK_HELP_ROLES: ReadonlySet<StaffRole> = new Set<StaffRole>([
  'manager', 'front_desk', 'maintenance', 'housekeeping', 'room_service',
]);

/** Roles that may browse the Help Desk sidebar. */
export const BROWSE_HELP_ROLES: ReadonlySet<StaffRole> = new Set<StaffRole>(['manager', 'front_desk']);

/** Roles that may manually reassign a task to another team. */
export const REASSIGN_TEAM_ROLES: ReadonlySet<StaffRole> = new Set<StaffRole>([
  'manager',
  'front_desk',
]);

/** Valid reassignment targets (HANDOFF is resolved at log time). */
export const REASSIGNABLE_ACTION_TYPES: ActionType[] = [
  'MAINTENANCE',
  'HOUSEKEEPING',
  'ROOM_SERVICE',
  'CONTACT_FRONT_DESK',
];

export function canReassignTaskTeam(role: StaffRole): boolean {
  return REASSIGN_TEAM_ROLES.has(role);
}

export function getAllowedNav(role: StaffRole): StaffNavId[] {
  return Array.from(ROLE_NAV[role]) as StaffNavId[];
}

export function getAllowedActionTypes(role: StaffRole): ActionType[] {
  return Array.from(ROLE_ACTION_TYPES[role]) as ActionType[];
}

File: backend/app/services/staff_permissions.py
"""RBAC permission matrix and FastAPI dependency for staff authentication."""
from __future__ import annotations

import hashlib
from dataclasses import dataclass
from typing import Callable, FrozenSet, Optional

from fastapi import Depends, Header, HTTPException

from app.core.config import get_settings
from app.models.schemas import ActionType, StaffMemberStatus, StaffRole
from app.services.database import get_database

# ---------------------------------------------------------------------------
# Permission matrices (single source of truth — frontend mirrors these)
# ---------------------------------------------------------------------------

ROLE_NAV: dict[StaffRole, FrozenSet[str]] = {
    StaffRole.MANAGER: frozenset({
        "tasks", "assigned", "schedule", "review", "guest-chat", "help-desk", "knowledge"
    }),
    StaffRole.FRONT_DESK: frozenset({
        "tasks", "assigned", "schedule", "review", "guest-chat", "help-desk", "knowledge"
    }),
    StaffRole.MAINTENANCE: frozenset({"tasks", "assigned", "schedule"}),
    StaffRole.HOUSEKEEPING: frozenset({"tasks", "assigned", "schedule"}),
    StaffRole.ROOM_SERVICE: frozenset({"tasks", "assigned", "schedule"}),
}

ROLE_ACTION_TYPES: dict[StaffRole, FrozenSet[ActionType]] = {
    StaffRole.MANAGER: frozenset(ActionType),
    StaffRole.FRONT_DESK: frozenset(ActionType),
    StaffRole.MAINTENANCE: frozenset({ActionType.MAINTENANCE, ActionType.HANDOFF}),
    StaffRole.HOUSEKEEPING: frozenset({ActionType.HOUSEKEEPING, ActionType.HANDOFF}),
    StaffRole.ROOM_SERVICE: frozenset({ActionType.ROOM_SERVICE, ActionType.HANDOFF}),
}

# Roles that may see the "Get help with this task" button (Agent 6)
TASK_HELP_ROLES: FrozenSet[StaffRole] = frozenset(StaffRole)

# Roles that may browse the Help Desk sidebar
BROWSE_HELP_ROLES: FrozenSet[StaffRole] = frozenset({
    StaffRole.MANAGER, StaffRole.FRONT_DESK
})

# Roles that may manually reassign a task to another team
REASSIGN_TEAM_ROLES: FrozenSet[StaffRole] = frozenset({
    StaffRole.MANAGER, StaffRole.FRONT_DESK
})

# Valid reassignment targets (HANDOFF is resolved at log time)
REASSIGNABLE_ACTION_TYPES: FrozenSet[ActionType] = frozenset({
    ActionType.MAINTENANCE,
    ActionType.HOUSEKEEPING,
    ActionType.ROOM_SERVICE,
    ActionType.CONTACT_FRONT_DESK,
})


# ---------------------------------------------------------------------------
# StaffContext — resolved identity carried through a request
# ---------------------------------------------------------------------------

@dataclass
class StaffContext:
    id: str
    display_name: str
    staff_code: str
    role: StaffRole
    property_id: str


# ---------------------------------------------------------------------------
# FastAPI dependency
# ---------------------------------------------------------------------------

def get_current_staff(
    x_staff_key: Optional[str] = Header(None, alias="X-Staff-Key"),
) -> StaffContext:
    """Resolve X-Staff-Key to a StaffContext; raise 401 on failure."""
    if not x_staff_key:
        raise HTTPException(status_code=401, detail="Missing X-Staff-Key header")

    settings = get_settings()

    # Legacy bootstrap: shared dev/property key → synthetic manager
    if x_staff_key == settings.staff_access_key:
        return StaffContext(
            id="legacy-staff",
            display_name="Staff",
            staff_code="LEGACY",
            role=StaffRole.MANAGER,
            property_id=settings.property_id,
        )

    # Per-user key: hash and look up approved staff member
    key_hash = hashlib.sha256(x_staff_key.encode()).hexdigest()
    db = get_database()
    member = db.get_staff_member_by_access_key_hash(key_hash)

    if member is None or member.status != StaffMemberStatus.APPROVED:
        raise HTTPException(status_code=401, detail="Invalid or missing staff key")

    role = member.approved_role or StaffRole.FRONT_DESK
    return StaffContext(
        id=member.id,
        display_name=member.display_name,
        staff_code=member.staff_code,
        role=role,
        property_id=member.property_id,
    )


def require_role(*roles: StaffRole) -> Callable[..., StaffContext]:
    """Return a FastAPI dependency that enforces one of the given roles."""
    role_set = frozenset(roles)

    def _dep(ctx: StaffContext = Depends(get_current_staff)) -> StaffContext:
        if ctx.role not in role_set:
            raise HTTPException(status_code=403, detail="Insufficient permissions")
        return ctx

    return _dep
